### ISSUE TITLE
Crash with 4.2.0rc3: Unrecoverable error while evaluating node

### ISSUE THREAD

**Issue by thii** [bazelbuild/bazel#13834]
Doesn't happen with `last_green` Bazel, so it must have been fixed somewhere.

https://buildkite.com/bazel/rules-swift-swift/builds/2741#19c25d8a-5530-4aff-8cca-d0998059ac84 (the log was for a build with 4.1.0 but I have reproduced this locally with 4.2.0rc3).

cc @katre

```
(08:35:41) FATAL: bazel crashed due to an internal error. Printing stack trace:
  | java.lang.RuntimeException: Unrecoverable error while evaluating node 'ConfiguredTargetKey{label=//examples/xplatform/hello_world:hello_world, config=BuildConfigurationValue.Key[58765066c537a2addebe865dea2674b9eb2786d3980bb8070f08dabcf685a2a8]}' (requested by nodes )
  | at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:563)
  | at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:398)
  | at java.base/java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(Unknown Source)
  | at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
  | at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
  | at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source)
  | at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
  | at java.base/java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)
  | Caused by: net.starlark.java.eval.Starlark$UncheckedEvalException: java.lang.IllegalArgumentException: hello_world.modulewrap.o (Starlark stack: [_swift_binary_impl@/workdir/swift/internal/swift_binary_test.bzl:343:61, _swift_linking_rule_impl@/workdir/swift/internal/swift_binary_test.bzl:249:50, register_link_binary_action@/workdir/swift/internal/linking.bzl:255:26, link@<builtin>])
  | at net.starlark.java.eval.Starlark.fastcall(Starlark.java:610)
  | at net.starlark.java.eval.Eval.evalCall(Eval.java:642)
  | at net.starlark.java.eval.Eval.eval(Eval.java:460)
  | at net.starlark.java.eval.Eval.execReturn(Eval.java:242)
  | at net.starlark.java.eval.Eval.exec(Eval.java:279)
  | at net.starlark.java.eval.Eval.execStatements(Eval.java:81)
  | at net.starlark.java.eval.Eval.execFunctionBody(Eval.java:65)
  | at net.starlark.java.eval.StarlarkFunction.fastcall(StarlarkFunction.java:160)
  | at net.starlark.java.eval.Starlark.fastcall(Starlark.java:606)
  | at net.starlark.java.eval.Eval.evalCall(Eval.java:642)
  | at net.starlark.java.eval.Eval.eval(Eval.java:460)
  | at net.starlark.java.eval.Eval.execAssignment(Eval.java:109)
  | at net.starlark.java.eval.Eval.exec(Eval.java:261)
  | at net.starlark.java.eval.Eval.execStatements(Eval.java:81)
  | at net.starlark.java.eval.Eval.execFunctionBody(Eval.java:65)
  | at net.starlark.java.eval.StarlarkFunction.fastcall(StarlarkFunction.java:160)
  | at net.starlark.java.eval.Starlark.fastcall(Starlark.java:606)
  | at net.starlark.java.eval.Eval.evalCall(Eval.java:642)
  | at net.starlark.java.eval.Eval.eval(Eval.java:460)
  | at net.starlark.java.eval.Eval.execAssignment(Eval.java:109)
  | at net.starlark.java.eval.Eval.exec(Eval.java:261)
  | at net.starlark.java.eval.Eval.execStatements(Eval.java:81)
  | at net.starlark.java.eval.Eval.execFunctionBody(Eval.java:65)
  | at net.starlark.java.eval.StarlarkFunction.fastcall(StarlarkFunction.java:160)
  | at net.starlark.java.eval.Starlark.fastcall(Starlark.java:606)
  | at com.google.devtools.build.lib.analysis.starlark.StarlarkRuleConfiguredTargetUtil.buildRule(StarlarkRuleConfiguredTargetUtil.java:130)
  | at com.google.devtools.build.lib.analysis.ConfiguredTargetFactory.createRule(ConfiguredTargetFactory.java:365)
  | at com.google.devtools.build.lib.analysis.ConfiguredTargetFactory.createConfiguredTarget(ConfiguredTargetFactory.java:194)
  | at com.google.devtools.build.lib.skyframe.SkyframeBuildView.createConfiguredTarget(SkyframeBuildView.java:938)
  | at com.google.devtools.build.lib.skyframe.ConfiguredTargetFunction.createConfiguredTarget(ConfiguredTargetFunction.java:1018)
  | at com.google.devtools.build.lib.skyframe.ConfiguredTargetFunction.compute(ConfiguredTargetFunction.java:371)
  | at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:477)
  | ... 7 more
  | Caused by: java.lang.IllegalArgumentException: hello_world.modulewrap.o
  | at com.google.common.base.Preconditions.checkArgument(Preconditions.java:142)
  | at com.google.devtools.build.lib.rules.cpp.CppLinkActionBuilder.addNonCodeInput(CppLinkActionBuilder.java:1334)
  | at com.google.devtools.build.lib.rules.cpp.CppLinkActionBuilder.addNonCodeInputs(CppLinkActionBuilder.java:1320)
  | at com.google.devtools.build.lib.rules.cpp.CppLinkActionBuilder.addLinkParams(CppLinkActionBuilder.java:1465)
  | at com.google.devtools.build.lib.rules.cpp.CcLinkingHelper.createDynamicLinkAction(CcLinkingHelper.java:739)
  | at com.google.devtools.build.lib.rules.cpp.CcLinkingHelper.createCcLinkActions(CcLinkingHelper.java:463)
  | at com.google.devtools.build.lib.rules.cpp.CcLinkingHelper.link(CcLinkingHelper.java:372)
  | at com.google.devtools.build.lib.rules.cpp.CcModule.link(CcModule.java:1989)
  | at com.google.devtools.build.lib.bazel.rules.cpp.BazelCcModule.link(BazelCcModule.java:140)
  | at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
  | at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
  | at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
  | at java.base/java.lang.reflect.Method.invoke(Unknown Source)
  | at net.starlark.java.eval.MethodDescriptor.call(MethodDescriptor.java:135)
  | at net.starlark.java.eval.BuiltinFunction.fastcall(BuiltinFunction.java:77)
  | at net.starlark.java.eval.Starlark.fastcall(Starlark.java:606)
  | ... 38 more
  | bazel build failed with exit code 37
 

<br class="Apple-interchange-newline" style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: normal; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none;">
```

### What operating system are you running Bazel on?

Linux.

**Comment by katre** [comment]
Just from checking the code, this appears to have been fixed because of 31b689ba7238abe5629a0de87fb6e245910526bb.

@oquenchil, can you take a look and let me know if I need to cherrypick that commit into 4.2.0?

**Comment by katre** [comment]
According to @oquenchil this commit should handle the problem and be safe to cherrypick. I'll see if this can be done today.

**Comment by thii** [comment]
Submitted here https://github.com/bazelbuild/bazel/pull/13838. There were two conflicts, hope I fixed them correctly.

**Comment by katre** [comment]
Fixed by #13838.
